File: frontend/lib/models/visitor.dart
class Visitor {
  final String id;
  final String tenantId;
  final String name;
  final String? phone;
  final String? email;
  final String? photoUrl;
  final String? idProofType;
  final String? idProofNumber;
  final String? company;
  final String? address;
  final DateTime createdAt;

  Visitor({
    required this.id,
    required this.tenantId,
    required this.name,
    this.phone,
    this.email,
    this.photoUrl,
    this.idProofType,
    this.idProofNumber,
    this.company,
    this.address,
    required this.createdAt,
  });

  factory Visitor.fromJson(Map<String, dynamic> json) {
    return Visitor(
      id: json['id'] as String,
      tenantId: json['tenant_id'] as String,
      name: json['name'] as String,
      phone: json['phone'] as String?,
      email: json['email'] as String?,
      photoUrl: json['photo_url'] as String?,
      idProofType: json['id_proof_type'] as String?,
      idProofNumber: json['id_proof_number'] as String?,
      company: json['company'] as String?,
      address: json['address'] as String?,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'tenant_id': tenantId,
      'name': name,
      'phone': phone,
      'email': email,
      'photo_url': photoUrl,
      'id_proof_type': idProofType,
      'id_proof_number': idProofNumber,
      'company': company,
      'address': address,
    };
  }
}

class VisitorPass {
  final String id;
  final String tenantId;
  final String visitorId;
  final String hostEmployeeId;
  final String purpose;
  final DateTime expectedDate;
  final DateTime? checkInTime;
  final DateTime? checkOutTime;
  final String passNumber;
  final String status;
  final String? badgeNumber;
  final String? zoneAccess;
  final bool visitorDeskValidated;
  final DateTime createdAt;
  final String? visitorName;
  final String? hostName;

  VisitorPass({
    required this.id,
    required this.tenantId,
    required this.visitorId,
    required this.hostEmployeeId,
    required this.purpose,
    required this.expectedDate,
    this.checkInTime,
    this.checkOutTime,
    required this.passNumber,
    required this.status,
    this.badgeNumber,
    this.zoneAccess,
    required this.visitorDeskValidated,
    required this.createdAt,
    this.visitorName,
    this.hostName,
  });

  factory VisitorPass.fromJson(Map<String, dynamic> json) {
    return VisitorPass(
      id: json['id'] as String,
      tenantId: json['tenant_id'] as String,
      visitorId: json['visitor_id'] as String,
      hostEmployeeId: json['host_employee_id'] as String,
      purpose: json['purpose'] as String,
      expectedDate: DateTime.parse(json['expected_date'] as String),
      checkInTime: json['check_in_time'] != null ? DateTime.parse(json['check_in_time'] as String) : null,
      checkOutTime: json['check_out_time'] != null ? DateTime.parse(json['check_out_time'] as String) : null,
      passNumber: json['pass_number'] as String,
      status: json['status'] as String? ?? 'scheduled',
      badgeNumber: json['badge_number'] as String?,
      zoneAccess: json['zone_access'] as String?,
      visitorDeskValidated: json['visitor_desk_validated'] as bool? ?? false,
      createdAt: DateTime.parse(json['created_at'] as String),
      visitorName: json['visitor_name'] as String?,
      hostName: json['host_name'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'tenant_id': tenantId,
      'visitor_id': visitorId,
      'host_employee_id': hostEmployeeId,
      'purpose': purpose,
      'expected_date': expectedDate.toIso8601String().substring(0, 10),
      'badge_number': badgeNumber,
      'zone_access': zoneAccess,
    };
  }
}

File: backend/app/schemas/visitor.py
"""Visitor schemas."""

import uuid
from datetime import datetime, date
from typing import Optional
from pydantic import BaseModel, Field, ConfigDict


class VisitorCreate(BaseModel):
    name: str = Field(..., max_length=200)
    phone: Optional[str] = Field(None, max_length=20)
    email: Optional[str] = None
    id_proof_type: Optional[str] = Field(None, max_length=50)
    id_proof_number: Optional[str] = Field(None, max_length=50)
    company: Optional[str] = Field(None, max_length=200)
    address: Optional[str] = None


class VisitorResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    tenant_id: uuid.UUID
    name: str
    phone: Optional[str] = None
    email: Optional[str] = None
    photo_url: Optional[str] = None
    id_proof_type: Optional[str] = None
    id_proof_number: Optional[str] = None
    company: Optional[str] = None
    address: Optional[str] = None
    created_at: datetime


class VisitorPassCreate(BaseModel):
    visitor_id: uuid.UUID
    host_employee_id: uuid.UUID
    purpose: str = Field(..., max_length=500)
    expected_date: date
    zone_access: Optional[str] = None


class VisitorPassResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    tenant_id: uuid.UUID
    visitor_id: uuid.UUID
    host_employee_id: uuid.UUID
    purpose: str
    expected_date: date
    check_in_time: Optional[datetime] = None
    check_out_time: Optional[datetime] = None
    pass_number: str
    status: str
    badge_number: Optional[str] = None
    zone_access: Optional[str] = None
    visitor_desk_validated: bool = False
    created_at: datetime
    visitor_name: Optional[str] = None
    host_name: Optional[str] = None


class VisitorCheckIn(BaseModel):
    pass_id: uuid.UUID


class VisitorCheckOut(BaseModel):
    pass_id: uuid.UUID


class VisitorFilter(BaseModel):
    from_date: Optional[date] = None
    to_date: Optional[date] = None
    status: Optional[str] = None
    host_employee_id: Optional[uuid.UUID] = None
    search: Optional[str] = None
